README.md
<<<<<<< HEAD
Privacy Badger [![Build Status](https://travis-ci.org/EFForg/privacybadgerchrome.svg)](https://travis-ci.org/EFForg/privacybadgerchrome)
===================
Privacy Badger blocks spying ads and invisible trackers. It's there to ensure that companies can't track your browsing without your consent.

This extension is designed to automatically protect your privacy from third party trackers that load invisibly when you browse the web. We send the Do Not Track header with each request, and our extension evaluates the likelihood that you are still being tracked. If the algorithm deems the likelihood is too high, we automatically block your browser from responding to the domain. Just because a domain has been flagged by Privacy Badger's algorithm, doesn't mean that that domain is tracking you, just that it could be. 

Our extension has three states. Red means Privacy Badger believes this third-party domain is a tracker, and has blocked it. Yellow means the domain is believed to be both a tracker and necessary for the functioning of the page, so Privacy Badger is allowing it but blocking its cookies to prevent it from uniquely identifying you. Green means that Privacy Badger believes this is not tracker. You can click on the Privacy Badger icon in your browser's toolbar if you wish to override the automatic blocking settings. Or, you can browse in peace as Privacy Badger starts finding and eating up web trackers one by one.

Nothing can stop the Privacy Badger from eating cookies when it's hungry!

Privacy Badger is a project of the Electronic Frontier Foundation.

##Developing
For an easy build, simply enable developer mode in chrome://extensions, hit
the "load unpacked extension" button and load up this directory.

Within the command line, install the dependencies.

```bash
$ npm install
```

### Testing

After "unpacking" the extension, find your extension's ID and
visit `chrome-extension://YOUR_EXTENSION_ID/tests/index.html`, replacing
`YOUR_EXTENSION_ID` with your 32 character ID.

For Selenium tests, run `./run_selenium_tests.sh` in the `tests` directory. 
You need to have `chromedriver`, `xvfb` and `python-virtualenv` installed.

This project is using the [QUnit](http://qunitjs.com/), [py.test](http://pytest.org/), [Selenium](http://www.seleniumhq.org/) test frameworks 
along with [Travis CI](https://travis-ci.org/) for continuous integration.

##License
Privacy Badger is licensed under the GPLv3. See LICENSE for more details
=======
# Privacy Badger Firefox [![Build Status](https://travis-ci.org/EFForg/privacybadgerfirefox.svg)](https://travis-ci.org/EFForg/privacybadgerfirefox)

## Introduction

Privacy Badger is a Firefox add-on that blocks spying ads and invisible trackers as you browse. [More info here.](https://www.eff.org/privacybadger)

## Developer guide

### Getting started
1. Install the `jpm` package using npm. `npm install -g jpm`
2.  Once the SDK is activated, run `jpm -b <path/to/firefox> run` to launch a clean Firefox profile with Privacy Badger installed. Run `jpm -b <path/to/firefox> test` to run tests. `jpm xpi` creates a package (.xpi file) that you can install by loading into Firefox.

### Important directories and files

    hooks/                    Git hooks. You can use them by copying into `.git/hooks`. The pre-push hook runs tests and cancels the push if they fail.

    release-utils/            | Files for making a self-hosted release of Privacy Badger and updates that are signed with an offline private key.
    Makefile                  | You probably don't need to worry about these unless you're a project maintainer.

    package.json              |
    data/                     |
    lib/                      | Most of the code that runs in the add-on. See SDK documentation for more info on the directory structure.
    test/                     |
    defaults/                 |

    doc/                      Changelog, style guide, how to make a signed release, other documentation TBD.

### Contributing

Before you submit a pull request please consult the [CONTRIBUTING.md](./CONTRIBUTING.md) file.

## How heuristic blocking works

This is a rough summary of Privacy Badger's internal logic for blocking trackers. At the moment, "tracker" == "third-party cookie from a site that tracks you on multiple first-party origins." I am in the process of adding support for other non-cookie tracker types (local storage, etags, cache hits, etc.).

Privacy Badger uses a (relatively-simple) heuristic algorithm for deciding whether a third-party is tracking you. When Privacy Badger sees a third-party request on a website, it checks:

1. Does the third-party read a cookie? If not, don't count it in the blocking heuristic. Otherwise:
2. Is the cookie sufficiently high-entropy? If not, don't count it. (Currently the entropy calculation is *very* crude! See lib/heuristicBlocker.js.) Otherwise:
3. Increment the heuristic blocker counter by +1 for that domain. Has the base domain (eTLD+1) of the third-party read cookies on at least 3 first-party base domains? If not, don't block it (for now). Otherwise:
4. Has the third party posted an acceptable DNT policy? (We check this using an XML HTTP Request to a well-known path where we are asking sites to post statements of [compliance with DNT](https://www.eff.org/dnt-policy).) If so, don't block it. Otherwise:
5. Is the third party or any of its parent domains on a preloaded whitelist of sites to not block because it would probably cause the first-party site to break? If so, block it from reading cookies in a third-party context. Otherwise:
6. Block third-party requests from the third-party entirely.

In addition, Privacy Badger will block third-party cookies from a domain if any of its parent domains have been blocked or cookie-blocked.

Note that users can manually set domains to be unblocked (green), cookie-blocked (yellow), or red (blocked). These choices *always* override the heuristic blocker.

By default, Privacy Badger sends the Do Not Track header on all requests. It also clears the referer for all requests that are cookie-blocked.

## Contact

The current maintainers of this project are  Cooper Quintin (cjq at eff dot org) and Noah Swartz (noah at eff dot org). There is also a [mailing list](https://lists.eff.org/mailman/listinfo/privacybadger) to discuss Privacy Badger development for both Firefox and Chrome.
>>>>>>> 062ef5088890cfa5967254a9732f81c4d9dcac00


doc/Changelog
1.9.2                             (2016-11-4)
  * Only run migration script on unload

1.9.1                             (2016-11-2)
  * Upgrade Jquery for mozilla validation
  * Only migrate data every 24 hours so as to not impact performance

1.9                               (2016-11-1)
  * Preparing for transition to web extensions
  * Changes to Chinese and Spanish translations

1.8				  (2016-8-8)
  * Updates to french and russian translation
  * Minor UI updates
  * fixes breaking bug in firefox 38

1.7				  (2016-3-6)
  * Bug fixes for options page
  * Translation Improvements
  * New donate button for EFF

1.0.6				  (2016-3-2)
  * New feature: Search within blocked domain list
  * New feature: Use wildcard in disabled sites list
  * Replace soundcloud widget with a click to play button
  * Misc. bug fixes and translation improvements

1.0.5				  (2015-12-3)
  * add spanish translation
  * add chinese translation
  * fix icon not displaying bug
  * badge counter displays number of domains blocked and cookieblocked instead
  of all potential trackers

1.0.3				  (2015-10-01)
  * Fix slow options page bug
  * Fix bug where privacy badger was blocking things it shouldn't when 
  cookies were disabled in firefox
  * Remove ABP code dependencies
  * Ukranian Localization
  * Improvements to the first run page 

1.0.2				  (2015-08-25)
  * Change id of self hosted addon for AMO signing, no other updates

1.0.1				  (2015-08-13)
  * Fixes bug where local storage being turned off causes privacy badger to
  fail
  * Fixes UI bug where two sliders are displayed in the popup
  * Fixes bug where, tutorial nagging box never goes away. 

1.0.0				  (2015-08-05)

  * Bugfixes from 0.99 
  * Detects Canvas Fingerprinting
  * Detect Local Storage Supercookies
  * Improved UI
  * Options page for overriding privacy badger settings
  * Report Broken Site button 
  * Many Bugfixes (see github)
  * Translations into swedish, french and german

0.99				  (2015-07-24)
  * Release Candidate for 1.0

0.2.6.1				  (2015-04-02)
  * Minor version bump to appease AMO, no other changes

0.2.6				  (2015-04-01)
  * Various security tweaks and bugfixes for AMO
  * Make tracker counting badger optional

0.2.5					  (2014-12-18)
  * Fixes issue with disqus 'allow domain' dialog infinite loop
  * Adds counter of potential trackers on a site for FF >= 36.0

0.2.4					  (2014-12-17)
  * Dialog to allow user to whitelist certain trackers to improve functionality.  
  * Fixes to icon display code to make Privacy Badger work in nightly.
  * Revamped icon that should look better on dark backgrounds.
  * Misc other bug fixes

0.2.3					  (2014-10-16)
  * Strip cookies from DNT policy recheck
  * Don't recheck DNT policy each time the browser is started
  * User preference to turn off DNT policy check

0.2.2					  (2014-09-16)
  * Fixes tests
  * Fixes console deprecation warning about widget library
  * Rechecks blocked sites for DNT Policy

0.2.1                                     (2014-07-21)
  * Migrate ShareMeNot icons to native icons with badger badge (#174)
  * Fix console errors due to empty object lookups in WeakMaps

0.2                                       (2014-07-17)
  * Incorporate ShareMeNot functionality.
  * Smoother UI for "reset to original state" panel button
  * Sort origins in UI by string-reversed alphabetical order
  * Show eTLD+1 of very long host names in the panel.
  * Attempt to not clear user-persistent cookies.
  * Inject cookies for requests in disabled windows.

0.1.4                                     (2014-06-05)
  * Support per-window private browsing (#11).
  * Fix bug caused by missing nsIProgressListener in additional windows.
  * Set DNT via prefs instead of directly modifying headers (#91).

0.1.3                                     (2014-05-12)
  * Make "unblock trackers" functionality clearer and add confirmation dialog
  (#80).
  * Duplicate "disable this site only" button in settings menu.
  * Check each parent domain for red/yellow in deciding whether to cookieblock
  (#81).
  * Operate consistently w/ 3rd party cookies blocked by default (#64).
  * Delete exempted cookies when browser closes if user has chosen "delete all
  cookies on browser close" in Firefox (#85).

0.1.2                                     (2014-05-05)
  * Fix button icon issues (#72).
  * Make third-party cookies warning clearer.

0.1.1                                     (2014-05-02)
  * Whitelists a domain if any of its parent domains are whitelisted (#58)
  * Update README.md.

0.1                                       (2014-05-01)
  * hard alpha release of privacy badger firefox.


lib/ui.js
// Privacy Badger user interface controller.

/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
//jshint moz:true
//jshint node:true
"use strict";

const { storage } = require("sdk/simple-storage");
const Request = require("sdk/request").Request;
const panel = require("sdk/panel");
const userStorage = require("./userStorage");
const utils = require("./utils");
const { on, once, off, emit } = require('sdk/event/core');
const main = require('./main');
const tabs = require("sdk/tabs");
const { getMostRecentBrowserWindow } = require("sdk/window/utils");
const { id: addonID, data } = require("sdk/self");
const contentPolicy = require('./contentPolicy');
var version = require("./package.json").version;
var preferences = require("sdk/simple-prefs").prefs;
const { Ci } = require("chrome");
var pageMod = require("sdk/page-mod");
const xulapp = require("sdk/system/xul-app");
const usingAustralis = xulapp.satisfiesVersion(">=29");
const pWidth = 385;
const pHeight = 560;

if (usingAustralis) {
  var { ToggleButton } = require("sdk/ui/button/toggle");
} else {
  const {
    Widget
  } = require("sdk/widget");
}
exports.usingAustralis = usingAustralis;
// Panel communicates with the content script (popup.js) using the port APIs.
// This is where the user toggles settings.

var pbPanel = panel.Panel({
  contentURL: data.url("popup.html"),
  contentScriptFile: [data.url("jquery-ui/js/jquery-2.2.4.min.js"),
                      data.url("jquery-ui/js/jquery-ui.min.js"),
                      data.url("tld.js"),
                      data.url("popup.js")],
  width: pWidth,
  height: pHeight,
  onShow: emitRefresh,
  onHide: function() {
    this.port.emit('hide');
    handleNewSettings(changedSettings);
    changedSettings = {};
    if (usingAustralis) {
      // Need to manually "un-toggle" the ToggleButton :/
      button.state('window', {checked: false});
    }
  }
});

// Refresh the panel
function emitRefresh() {
  if (userStorage.isDisabledSite(tabs.activeTab.url, tabs.activeTab)) {
    pbPanel.port.emit("show-inactive");
    return;
  }
  let settings = getCurrentSettings();
  let seenComic = userStorage.getSeenComic();
  console.log("Showing panel with settings: "+JSON.stringify(settings));
  pbPanel.port.emit("show-trackers", settings, storage, seenComic);
}

// Called when user reports error
pbPanel.port.on("report", function(blob) {
  var url = tabs.activeTab.url;
  blob["url"] = url;
  blob["version"] = version;
  blob["fqdn"] = url.split("/",3)[2];
  var report = Request({
    url: "https://privacybadger.org/reporting",
    content: JSON.stringify(blob),
    contentType: "application/json",
    onComplete: function(response) {
        if(response.status == 200){
          pbPanel.port.emit("report-success");
        } else {
          pbPanel.port.emit("report-fail");
        }
    }
  });
  report.post();
});

// Called whenever user toggles a setting
pbPanel.port.on("update", function(data) {
  handleUpdate(data);
});
// Called when user setting is undone
pbPanel.port.on("reset", function(origin) {
  changedSettings[origin] = "reset";
  let originalState = contentPolicy.Policy.getOriginalState(origin);
  pbPanel.port.emit("change-setting", { origin: origin, action: originalState, flag: false });
});
// Activate PB on the current page
pbPanel.port.on("activateSite", function() {
  let currentTab = tabs.activeTab;
  let host = require("sdk/url").URL(currentTab.url).host;
  userStorage.removeFromDisabledSites(host, currentTab);
  let topContentWindow = utils.getMostRecentContentWindow();
  if (!topContentWindow) { return; }
  settingsMap.set(topContentWindow, {cleared: true});
  tabs.activeTab.reload();
  emitRefresh();
});
pbPanel.port.on("openSlideshow", function() {
  tabs.open(data.url("firstRun.html#slideshow"));
  pbPanel.hide();
});
pbPanel.port.on("openOptions", function(){
  tabs.open(data.url("options.html"));
  pbPanel.hide();
});
pbPanel.port.on("openDonate", function(){
  tabs.open("https://supporters.eff.org/donate/support-privacy-badger");
  pbPanel.hide();
});
pbPanel.port.on("openHelp", function(){
  tabs.open(data.url("firstRun.html")); 
  pbPanel.hide();
});
// Deactivate PB on the current page
pbPanel.port.on("deactivateSite", function() {
  let currentTab = tabs.activeTab;
  let host = require("sdk/url").URL(currentTab.url).host;
  userStorage.addToDisabledSites(host, currentTab);
  
  let topContentWindow = utils.getMostRecentContentWindow();
  if (!topContentWindow) { return; }
  settingsMap.set(topContentWindow, {cleared: true});
  tabs.activeTab.reload();
  emitRefresh();
});
// Unblock everything
pbPanel.port.on("unblockAll", function() {
  main.emptyData();
  clearSettings();
  utils.reloadCurrentTab();
  emitRefresh();
});
// Close the panel
pbPanel.port.on("hidePanel", function() {
  pbPanel.hide();
});

// Begin hack to attach the panel to the ToggleButton in 29
const buttonPrefix =
  'button--' + addonID.toLowerCase().replace(/[^a-z0-9-_]/g, '');

const toWidgetID = id => buttonPrefix + '-' + id;

const nodeFor = ({id}) =>
  getMostRecentBrowserWindow().document.getElementById(toWidgetID(id));
// end hack

// This is the little button in the addons bar that opens the panel on click.
let button;
if (usingAustralis) {
  button = ToggleButton({
    id: "pb-button",
    label: "Privacy Badger",
    icon: {
      "16": data.url("icons/badger-16.png"),
      "32": data.url("icons/badger-32.png"),
      "64": data.url("icons/badger-64.png")
    },
    onChange: function(state) {
      if (state.checked) {
        // in FF29 will show a warning about using the 2nd argument,
        // but only in dev mode
        pbPanel.show({ position: button }, nodeFor(button));
      }
    }
  });
} else {
  button = Widget({
    id: "pb-button",
    label: "Privacy Badger",
    contentURL: data.url("icons/badger-64.png"),
    panel: pbPanel
  });
}

exports.pbPanel = pbPanel;
exports.pbButton = button;

/**
 * Update tab's action badge with current count based on channel
 */
function updateCountForChannel(channel, aWin) {
  if(!channel) { return {}; }

  if(arguments.length < 2) {
    aWin = utils.getTopWindowForChannel(channel);
  }

  if(!aWin) { return {}; }

  let tab = utils.getTabForChannel(channel, aWin);

  if(contentPolicy.Policy.isDisabledRequest(channel, aWin)) {
    if(xulapp.satisfiesVersion(">=36") && tab) {
      button.state(tab, {
        icon: {
          "16": data.url("icons/badger-grey-16.png"),
          "32": data.url("icons/badger-grey-32.png"),
          "64": data.url("icons/badger-grey-64.png")
        },
      });
    }
    return;
  }

  var settings = getSettingsForWindow(aWin);
  var numBlocked = 0;
  for (var origin in settings){
    let action = settings[origin];
    if (['block', 'userblock', 'cookieblock', 'usercookieblock'].indexOf(action) > -1) {
      numBlocked++;
    }
  }
  var badgeText = numBlocked + "";
  var badgeColor = "#FF3636";
  if(numBlocked == 0){
    badgeColor = "#77DD77";
  }
  if(xulapp.satisfiesVersion(">=36") && tab) {
    // Note that you must set both properties simultaneously, otherwise
    // the one you don't set will get overwritten with the default value
    button.state(tab, {badge: badgeText, badgeColor: badgeColor});
  } 
}
exports.updateCountForChannel = updateCountForChannel;

/**
 * Retrieve and update block settings based on user interactions.
 *
 *  getCurrentSettings - returns a "dictionary" with the third-party tracking
 *  origins as keys and one of the following as values:
 *    * block: origin has been blocked by heuristic
 *    * userblock: origin has been blocked by user
 *    * cookieblock: origin is cookie-blocked by heuristics
 *    * usernoaction: origin has been whitelisted by user
 *    * usercookieblock: origin is cookie-blocked by user
 *    * noaction: none of the above
 *
 *  handleNewSettings - takes as input ONLY the settings that were changed
 *  after the user interacted with the popup, based on the "userset"
 *  attribute. updates Storage accordingly.
 *
 *  handleUpdate - stores the changed setting in a temporary object.
 *
 */

// origin-action keypairs
let changedSettings = {};

// Special key "cleared" in settings marks that a page needs to be reloaded
// before its settings are shown correctly
function clearSettings() {
  tempUnblockMap.clear();
  settingsMap.clear();
  var allWindows = utils.getAllWindows();
  console.log("ALL WINDOWS", allWindows);
  allWindows.forEach(function(element, index, array) {
    settingsMap.set(element, {cleared: true});
  });
}

function getCurrentSettings() {
  let topContentWindow = utils.getMostRecentContentWindow();
  console.log("LOCATION topContentWindow", topContentWindow.location.href);
  if (!topContentWindow) { return {}; }
  return settingsMap.get(topContentWindow) || {};
}

function getSettingsForWindow(aWin) {
  if(!aWin) { return {}; }
  return settingsMap.get(aWin) || {};
}
exports.getSettingsForWindow = getSettingsForWindow;

function handleUpdate(data) {
  changedSettings[data.origin] = data.action;
}

function handleNewSettings(settings, silent) {
  if (Object.keys(settings).length === 0) { return false; }
  console.log("handling new settings", JSON.stringify(settings));
  for (let origin in settings) {
    switch (settings[origin]) {
      case "reset":
        let topContentWindow = utils.getMostRecentContentWindow();
        userStorage.resetOrigin(origin, topContentWindow);
        break;
      case "block":
        userStorage.add("red", origin);
        break;
      case "cookieblock":
        userStorage.add("yellow", origin);
        break;
      case "noaction":
        userStorage.add("green", origin);
    }
  }
  if(!silent){
    utils.reloadCurrentTab();
  }
  pbPanel.port.emit("give-storage", storage);
  return true;
}

/**
 * settingsMap implementation. This object keeps track of which allow/block
 * settings were applied to each window.
 */
let settingsMap = new WeakMap();

/**
 * tempUnblockMap keeps track of temporary unblocks due to social media widget clicks
 * (should eventually refactor into a property of settingsMap)
 */
let tempUnblockMap = new WeakMap();

/**
 * tabWorkers keeps track of the workers for open tabs, so we can send messages
 * to their content scripts when blocking happens.
 */
let tabWorkers = new WeakMap();

/**
 * Register our event listener to update Settings. Events are emitted
 * in the form emit(target, type, msg, nsIDOMWindow, origin),
 * where origin is the cookie origin and nsIDOMWindow is the parent origin.
 */
on(settingsMap, "update-settings", updateSettingsListener);

/**
 * Ways for settingsMap value to get updated for a domwin:
 *   third party cookie is set = noaction
 *   third party cookie on preloads list is heuristic-blacklisted: cookieblock
 *   third party cookie is heuristics-blacklisted = block
 *   user sets green on third party cookie = usernoaction
 *   user sets yellow on third party cookie = usercookieblock
 *   user sets red on third party cookie = userblock
 *
 *   @param {string} msg
 *   @param {nsIDOMWindow} aWin
 *   @param {string} aOrigin
 *   @return {null}
 */
function updateSettingsListener(msg, aWin, aOrigin) {
  if (!aWin) {
    console.log("Can't update request without a window");
    return;
  } else if (!aOrigin) {
    console.log("Missing origin for cookie");
    return;
  } else if (!(aWin instanceof Ci.nsIDOMWindow)){
    console.log("ERROR settings key not a dom window object");
    return;
  }
  aWin = aWin.top;
  var setting = settingsMap.get(aWin) || {};
  setting[aOrigin] = msg;

  settingsMap.set(aWin, setting);

  notifyContentScript(msg, aWin, aOrigin);
}

/**
 * Notify the tab's content script so that it can replace social widgets,
 * if applicable.
 */
function notifyContentScript(msg, aWin, aOrigin) {
  if (msg == "block" || msg == "userblock") {
    let tabWorker = tabWorkers.get(aWin);
    if (tabWorker && tabWorker.tab) {
      tabWorker.port.emit("replaceSocialWidget", aOrigin);
    }
  }
}

/**
 * Used on location change.
 *
 * @param {nsIDOMWindow} aWin
 */
exports.clearSettingsMap = function(aWin) {
  if (!aWin) { return; }
  settingsMap.delete(aWin);
};

/**
 * Handle firstrun page
 */
pageMod.PageMod({
  include: data.url("firstRun.html*"), // The * at the end is to detect url fragments
  contentScriptFile: [data.url("jquery-ui/js/jquery-2.2.4.min.js"),
                      data.url("firstRunCS.js")],
  contentScriptWhen: 'end',
  onAttach: function(worker) {
    worker.port.on('setSeenComic', function(setting){
      userStorage.setSeenComic(setting);
    });
  },
});

/**
 * Handle options page
 */
pageMod.PageMod({
  include: data.url("options.html"),
  contentScriptFile: [data.url("jquery-ui/js/jquery-2.2.4.min.js"),
                      data.url("jquery-ui/js/jquery-ui.min.js"),
                      data.url("options.js")],
  contentStyleFile: [data.url("skin/jquery-ui.css"),
                     data.url("skin/toggle-switch.css")],
  onAttach: function(worker){
    worker.port.on('reqSettings', function(){
      emitRcv(worker);
    });
    worker.port.on('removeFromDisabledSites', function(domains){
      for(let i = 0; i < domains.length; i++){
        console.log('domains', domains);
        console.log('remove', domains[i]);
        userStorage.removeFromDisabledSites(domains[i]);
      }
      emitRcv(worker);
    });
    worker.port.on('addToDisabledSites', function(host){
      userStorage.addToDisabledSites(host);
      emitRcv(worker);
    });
    worker.port.on('updateUserPref', function(pref){
      console.log('GOT PREF', pref);
      preferences[pref.name] = pref.value;
      emitRcv(worker);
    });
    worker.port.on("updateDomain", function(data) {
      handleUpdate(data);
      handleNewSettings(changedSettings, true);
      changedSettings = {};
      emitRcv(worker);
    });
    worker.port.on("resetDomain", function(origin) {
      changedSettings[origin] = "reset";
      handleNewSettings(changedSettings, true);
      changedSettings = {};
      emitRcv(worker);
    });
  },
    
});

function emitRcv(worker){
  var storage = userStorage.getAll();
  worker.port.emit('recvSettings', {
    disabledSites: storage.disabledSites, 
    prefs: preferences, 
    origins: loadOrigins(storage)
  });
}

function loadOrigins(storage){
  var origins = {};
  var origin = '';
  for(origin in storage.originFrequency){
    // If we have seen an origin it will at least be 'no action'
    origins[origin] = 'noaction';
    // If it is in blockedOrigins it will at least be blocked
    if(!!storage.blockedOrigins[origin]){
      origins[origin] = 'block';
    }
  } 

  for(origin in storage.preloads){
    let url = 'https://' + origin + '/';
    let baseDomain = utils.getBaseDomain(utils.makeURI(url));
    if(!!storage.blockedOrigins[baseDomain]){
      origins[origin] = 'cookieblock';
    }
  }

  for(origin in storage.userGreen ){
    origins[origin] = 'usernoaction';
  }
  for(origin in storage.userYellow){
    origins[origin] = 'usercookieblock';
  }
  for(origin in storage.userRed ){
    origins[origin] = 'userblock';
  }
  return origins;
}

//on(settingsMap, "clear-settings", clearSettingsListener);

exports.settingsMap = settingsMap;
exports.tempUnblockMap = tempUnblockMap;
exports.tabWorkers = tabWorkers;
